File: frontend/src/services/couponService.js
import axios from 'axios'

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8081'

const LOCAL_COUPONS = {
  WELCOME50: { code: 'WELCOME50', discountType: 'FIXED_AMOUNT', discountValue: 50000, minOrderValue: 0, usageLimit: 100, usedCount: 0, expiryDate: '2026-12-31 23:59:59', isActive: true },
  GIAM10: { code: 'GIAM10', discountType: 'PERCENTAGE', discountValue: 10, minOrderValue: 0, usageLimit: 100, usedCount: 5, expiryDate: '2026-12-31 23:59:59', isActive: true },
  TRU50K: { code: 'TRU50K', discountType: 'FIXED_AMOUNT', discountValue: 50000, minOrderValue: 2000000, usageLimit: 50, usedCount: 10, expiryDate: '2026-12-31 23:59:59', isActive: true },
  HETHAN: { code: 'HETHAN', discountType: 'PERCENTAGE', discountValue: 20, minOrderValue: 0, usageLimit: 100, usedCount: 0, expiryDate: '2023-01-01 00:00:00', isActive: true },
  HETLUOT: { code: 'HETLUOT', discountType: 'FIXED_AMOUNT', discountValue: 100000, minOrderValue: 0, usageLimit: 10, usedCount: 10, expiryDate: '2026-12-31 23:59:59', isActive: true },
}

function normalizeCode(code) {
  return String(code || '').trim().toUpperCase()
}

function calculateDiscount(coupon, subtotal) {
  const normalizedSubtotal = Math.max(0, Number(subtotal) || 0)
  const rawValue = Number(coupon?.discountValue) || 0
  const discountType = String(coupon?.discountType || '').trim().toUpperCase()

  if (discountType === 'PERCENTAGE') {
    return Math.min(normalizedSubtotal, Math.max(0, normalizedSubtotal * (rawValue / 100)))
  }

  if (discountType === 'FIXED_AMOUNT') {
    return Math.min(normalizedSubtotal, Math.max(0, rawValue))
  }

  return 0
}

function fallbackValidate(code, subtotal, shippingFee = 0) {
  const normalizedCode = normalizeCode(code)
  const coupon = LOCAL_COUPONS[normalizedCode]
  const normalizedSubtotal = Math.max(0, Number(subtotal) || 0)
  const normalizedShipping = Math.max(0, Number(shippingFee) || 0)

  const fail = (message) => ({
    success: false,
    message,
    subtotal: normalizedSubtotal,
    shippingFee: normalizedShipping,
    discountAmount: 0,
    total: normalizedSubtotal + normalizedShipping,
  })

  if (!coupon) return fail('Mã giảm giá không tồn tại')
  if (!coupon.isActive) return fail('Mã giảm giá đã bị khóa')
  if ((coupon.usedCount || 0) >= (coupon.usageLimit || 1)) return fail('Mã giảm giá đã hết lượt sử dụng')

  const expiry = new Date(coupon.expiryDate.replace(' ', 'T'))
  if (!Number.isNaN(expiry.getTime()) && expiry.getTime() <= Date.now()) return fail('Mã giảm giá đã hết hạn')
  if (normalizedSubtotal < Number(coupon.minOrderValue || 0)) return fail('Đơn hàng chưa đạt giá trị tối thiểu để áp mã giảm giá')

  const discountAmount = calculateDiscount(coupon, normalizedSubtotal)
  return {
    success: true,
    message: 'Áp mã giảm giá thành công',
    coupon,
    subtotal: normalizedSubtotal,
    shippingFee: normalizedShipping,
    discountAmount,
    total: Math.max(0, normalizedSubtotal + normalizedShipping - discountAmount),
  }
}

export async function validateCoupon(code, subtotal, shippingFee = 0) {
  if (import.meta.env.MODE === 'test') {
    return fallbackValidate(code, subtotal, shippingFee)
  }

  try {
    const response = await axios.post(`${API_BASE_URL}/api/coupons/validate`, {
      code,
      subtotal,
      shippingFee,
    })
    return response.data
  } catch (error) {
    if (error.response && error.response.data) {
      return error.response.data
    }

    return fallbackValidate(code, subtotal, shippingFee)
  }
}

File: backend/src/main/java/com/shopcart/service/CouponService.java
package com.shopcart.service;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;

import com.shopcart.dto.CouponValidationRequest;
import com.shopcart.dto.CouponValidationResponse;
import com.shopcart.entity.Coupon;
import com.shopcart.repository.CouponRepository;

@Service
public class CouponService {
    private static final DateTimeFormatter DATE_TIME_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");

    @Autowired
    private CouponRepository couponRepository;

    public CouponValidationResponse validateCoupon(CouponValidationRequest request) {
        BigDecimal subtotal = normalizeAmount(request != null ? request.getSubtotal() : null);
        BigDecimal shippingFee = normalizeAmount(request != null ? request.getShippingFee() : null);
        String code = request != null ? request.getCode() : null;

        if (!StringUtils.hasText(code)) {
            return failure("Vui lòng nhập mã giảm giá hợp lệ", subtotal, shippingFee);
        }

        Coupon coupon = couponRepository.findByCodeIgnoreCase(code.trim())
            .orElse(null);

        if (coupon == null) {
            return failure("Mã giảm giá không tồn tại", subtotal, shippingFee);
        }

        String validationError = validateCouponState(coupon, subtotal);
        if (validationError != null) {
            return failure(validationError, subtotal, shippingFee);
        }

        BigDecimal discount = calculateDiscount(coupon, subtotal);
        BigDecimal total = subtotal.add(shippingFee).subtract(discount);

        return CouponValidationResponse.builder()
            .success(true)
            .message("Áp mã giảm giá thành công")
            .coupon(toCouponInfo(coupon))
            .subtotal(subtotal)
            .shippingFee(shippingFee)
            .discountAmount(discount)
            .total(total.max(BigDecimal.ZERO))
            .build();
    }

    public Coupon requireValidCoupon(String code, BigDecimal subtotal) {
        if (!StringUtils.hasText(code)) {
            return null;
        }

        Coupon coupon = couponRepository.findByCodeIgnoreCase(code.trim())
            .orElseThrow(() -> new RuntimeException("Mã giảm giá không tồn tại"));

        String validationError = validateCouponState(coupon, normalizeAmount(subtotal));
        if (validationError != null) {
            throw new RuntimeException(validationError);
        }

        return coupon;
    }

    public BigDecimal calculateDiscount(Coupon coupon, BigDecimal subtotal) {
        if (coupon == null) {
            return BigDecimal.ZERO;
        }

        BigDecimal normalizedSubtotal = normalizeAmount(subtotal);
        String discountType = coupon.getDiscountType() == null ? "" : coupon.getDiscountType().trim().toUpperCase();
        BigDecimal discount = BigDecimal.ZERO;

        if ("PERCENTAGE".equals(discountType)) {
            discount = normalizedSubtotal.multiply(coupon.getDiscountValue()).divide(BigDecimal.valueOf(100), 2, RoundingMode.HALF_UP);
        } else if ("FIXED_AMOUNT".equals(discountType)) {
            discount = coupon.getDiscountValue();
        }

        return discount.max(BigDecimal.ZERO).min(normalizedSubtotal);
    }

    public void incrementUsage(Coupon coupon) {
        if (coupon == null) {
            return;
        }

        coupon.setUsedCount((coupon.getUsedCount() == null ? 0 : coupon.getUsedCount()) + 1);
        couponRepository.save(coupon);
    }

    private CouponValidationResponse failure(String message, BigDecimal subtotal, BigDecimal shippingFee) {
        return CouponValidationResponse.builder()
            .success(false)
            .message(message)
            .subtotal(subtotal)
            .shippingFee(shippingFee)
            .discountAmount(BigDecimal.ZERO)
            .total(subtotal.add(shippingFee))
            .build();
    }

    private CouponValidationResponse.CouponInfo toCouponInfo(Coupon coupon) {
        return CouponValidationResponse.CouponInfo.builder()
            .code(coupon.getCode())
            .discountType(coupon.getDiscountType())
            .discountValue(coupon.getDiscountValue())
            .minOrderValue(coupon.getMinOrderValue())
            .usageLimit(coupon.getUsageLimit())
            .usedCount(coupon.getUsedCount())
            .expiryDate(coupon.getExpiryDate())
            .isActive(coupon.getIsActive())
            .build();
    }

    private String validateCouponState(Coupon coupon, BigDecimal subtotal) {
        if (coupon.getIsActive() == null || !coupon.getIsActive()) {
            return "Mã giảm giá đã bị khóa";
        }

        if (coupon.getUsageLimit() != null && coupon.getUsedCount() != null && coupon.getUsedCount() >= coupon.getUsageLimit()) {
            return "Mã giảm giá đã hết lượt sử dụng";
        }

        LocalDateTime expiryDate = parseDateTime(coupon.getExpiryDate());
        if (expiryDate != null && !LocalDateTime.now().isBefore(expiryDate)) {
            return "Mã giảm giá đã hết hạn";
        }

        BigDecimal minOrderValue = normalizeAmount(coupon.getMinOrderValue());
        if (subtotal.compareTo(minOrderValue) < 0) {
            return "Đơn hàng chưa đạt giá trị tối thiểu để áp mã giảm giá";
        }

        return null;
    }

    private LocalDateTime parseDateTime(String value) {
        if (!StringUtils.hasText(value)) {
            return null;
        }

        try {
            return LocalDateTime.parse(value.trim(), DATE_TIME_FORMATTER);
        } catch (Exception _error) {
            return null;
        }
    }

    private BigDecimal normalizeAmount(BigDecimal value) {
        return value == null ? BigDecimal.ZERO : value.max(BigDecimal.ZERO);
    }
}